base/scopedvalues.jl

```diff
@@ -0,0 +1,175 @@
+#   the value portion is read-only/write-once, but the cache version
+#   would need a lock which makes ImmutableDict incredibly attractive.
+#   We could also use task-local-storage, but that added about 12ns.
+# - Values are GC'd when scopes become unreachable, one could use
+#   a WeakKeyDict to also ensure that values get GC'd when ScopedValues
+#   become unreachable.
+# - Scopes are an inline implementation of an ImmutableDict, if we wanted
+#   be really fancy we could use a CTrie or HAMT.
+
+mutable struct Scope
+    const parent::Union{Nothing, Scope}
+    const key::ScopedValue
+    const value::Any
+    Scope(parent, key::ScopedValue{T}, value::T) where T = new(parent, key, value)
+end
+Scope(parent, key::ScopedValue{T}, value) where T =
+    Scope(parent, key, convert(T, value))
+
+function Scope(scope, pairs::Pair{<:ScopedValue}...)
+    for pair in pairs
+        scope = Scope(scope, pair...)
+    end
+    return scope
+end
+
+"""
+    current_scope()::Union{Nothing, Scope}
+
+Return the current dynamic scope.
+"""
+current_scope() = current_task().scope::Union{Nothing, Scope}
+
+function Base.show(io::IO, scope::Scope)
+    print(io, Scope, "(")
+    seen = Set{ScopedValue}()
+    while scope !== nothing
+        if scope.key ∉ seen
+            if !isempty(seen)
+                print(io, ", ")
+            end
+            print(io, typeof(scope.key), "@")
+            show(io, Base.objectid(scope.key))
+            print(io, " => ")
+            show(IOContext(io, :typeinfo => eltype(scope.key)), scope.value)
+            push!(seen, scope.key)
+        end
+        scope = scope.parent
+    end
+    print(io, ")")
+end
+
+function Base.getindex(var::ScopedValue{T})::T where T
+    scope = current_scope()
+    while scope !== nothing
+        if scope.key === var
+            return scope.value::T
+        end
+        scope = scope.parent
+    end
+    return var.initial_value
+end
+
+function Base.show(io::IO, var::ScopedValue)
+    print(io, ScopedValue)
+    print(io, '{', eltype(var), '}')
+    print(io, '(')
+    show(IOContext(io, :typeinfo => eltype(var)), var[])
+    print(io, ')')
+end
+
+"""
+    scoped(f, (var::ScopedValue{T} => val::T)...)
+
+Execute `f` in a new scope with `var` set to `val`.
+"""
+function scoped(f, pair::Pair{<:ScopedValue}, rest::Pair{<:ScopedValue}...)
+    @nospecialize
+    ct = Base.current_task()
+    current_scope = ct.scope::Union{Nothing, Scope}
+    scope = Scope(current_scope, pair, rest...)
+    ct.scope = scope
+    try
+        return f()
+    finally
+        ct.scope = current_scope
+    end
+end
+
+scoped(@nospecialize(f)) = f()
+
+"""
+    @scoped vars... expr
+
+Macro version of `scoped(f, vars...)` but with `expr` instead of `f` function.
+This is similar to using [`scoped`](@ref) with a `do` block, but avoids creating
+a closure.
+"""
+macro scoped(exprs...)
+    ex = last(exprs)
+    if length(exprs) > 1
+        exprs = exprs[1:end-1]
+    else
+        exprs = ()
+    end
```

Comment:
should we check `exprs` isn't empty first, so we can throw something more informative than a BoundsError if it is?
```suggestion
    if isempty(exprs) 
        error("@scoped expects at least one argument")
    elseif length(exprs) == 1
        ex = only(args)
        exprs = ()
    else
        ex = last(exprs)
        exprs = exprs[1:end-1]
    end
```